### ISSUE TITLE
HidGuardian white-list cleanup.

### ISSUE THREAD

**Issue by JoshWobbles** [nefarius/ViGEm#27]
There seems to be a need for some sort of manual cleanup call to empty out whitelist applications. Perhaps this should even be done automatically on service start, and also would not be a bad idea if on an applications attempt to open a device, when the reg is checked for whitelist, that the driver goes through all the listed processes to see if they are still valid then remove if not.

**Comment by nefarius** [comment]
Yeah, current approach is giving developers too much freedom and I know what that means 😉

**Comment by JoshWobbles** [comment]
Part of the issue with the "orphaned" processes in there seems to stem from the fact that the pipe communication would sometimes break down between the UM library and the service and not reestablish, thus ignoring the call to de-white-list.

Have you thought of just moving that reg ket to an unprotected part of the reg thus eliminating need for external libraries and services altogether?

**Comment by nefarius** [comment]
To be honest the service/library implementation was just a quick brain fart of mine, I should really learn to stop pushing every stupid idea to GitHub 😆 

But seriously; drivers can only read a restricted portion of the registry and certainly no parts that are non-elevated accessible. I chose the registry keys for control since the driver has to know which device to block on startup, later blocking "while running" is not possible and could probably cause system instability.

TL;DR: I'm all open for better solutions.

**Comment by JoshWobbles** [comment]
Hmmm, can non elevated applications communicate directly to the driver via a named pipe?

**Comment by nefarius** [comment]
Unfortunately there is no such thing as named pipes or other well-known methods of IPC for drivers.

The main issue here is, that I was lazy enough to write the user-mode library as a managed DLL. Therefore I can not abuse the DllMain() function to execute code on process exit which could handle the clean-up in a safe way. I think I have to ditch the .NET approach here and rewrite it in native code.

**Comment by JoshWobbles** [comment]
Pretty sure shared memory works at the driver level

**Comment by nefarius** [comment]
[There actually is](http://www.osronline.com/article.cfm?article=39) but I have to read in detail if it's worth 😉 Feeling BSODs lurking in the dark...

**Comment by evilC** [comment]
If it is a choice between having to run an elevated process, and having no functionality in this area for an extended period of time while you find a better solution, I would take the elevated process in a heartbeat.  
Also, the HC.Srv web interface appears to have a button to remove all PIDs, so surely this is only a minor problem?  
HG has the potential to be a major game-changer, it seems like such a crime for it to sit unused over such a minor fly in the ointment with HC. Is it not best to triage it as best we can so that people can test HG and make sure it behaves as expected on a variety of systems?

**Comment by JoshWobbles** [comment]
@evilC  Could write your own service, it's only a band-aid solution but works ok. https://forums.vigem.org/d/13-hidguardian-whitelist-cleanup

**Comment by nefarius** [comment]
@evilC I don't see your point here. The basic idea was to implement automatic cleanup if a process dies and doesn't deregister itself so it's orphaned PID stays whitelisted. Since there is no reliable way in Windows to catch a dying process I can't really do much about this. This also has nothing to do with elevated privileges.

Having a developer force to gain elevated rights in his user-mode process is a bad idea. That's why the service exists. It abstracts ways access to the drivers registry settings via API and a small helper DLL.

You can request to empty the whitelist any time by calling http://localhost:26762/api/v1/hidguardian/whitelist/purge

**Comment by evilC** [comment]
Well I only this second figured out how to add a whitelisted process by looking at the [HC.Lib source](https://github.com/nefarius/ViGEm/blob/master/HidCerberus.Lib/HidCerberus.Lib.cpp#L45), so 
I have not had a chance to experience this issue first-hand.  

@JoshWobbles - that looks to be basically the idea I had in mind - thanks!

**Comment by Wolfereign** [comment]
Is it possible to base the whitelist on something other than the PID?  For example specific a exe name or exe full path base whitelist. 

C:\Games\UCR\ucr.exe

This way a permanent whitelist can be setup instead of a constantly revolving one.
